Add doc references

EmilySeville7cfg opened the pull request.

DannyBen reviewed (changes requested):
Two thoughts:

1. Isn't linking to https://madness.dannyb.co/#configuration-file better?
2. In cases like this, where all the links are the same, is it common practice to repeat it like this instead of just placing it once at the parent/top level?

EmilySeville7cfg commented:
> Isn't linking to https://madness.dannyb.co/#configuration-file better?

Yes, it is.

> In cases like this, where all the links are the same, is it common practice to repeat it like this instead of just placing it once at the parent/top level?

AFAIK there is no any best/common practice in general for this issue. When I introduce changes I orient on YAML Red Hat extension for VS Code, how it works. If I put link just at the top level property, it will be shown just here and not for nested properties.

EmilySeville7cfg commented:
@DannyBen, here is a nu script to generate JSON schema template from sample madness.yml file:

```nushell
let contents = "# .madness.yml

# path to the documentation root
path: .

# server port
port: 3000

# server listen address
bind: 0.0.0.0

# set a server root path, for example:
# base_uri: /docs
base_uri: ~

# enable sidebar
sidebar: true

# add H1 title to files that do not have one
auto_h1: true

# append navigation to directory READMEs
auto_nav: true

# replace <!-- TOC --> in any file with its internal table of contents
# set to true to enable it with the default '## Table of Contents' caption,
# or set to any string that will be inserted before it as a caption.
auto_toc: true

# enable syntax highlighter for code snippets
highlighter: true

# enable the copy to clipboard icon for code snippets
copy_code: true

# convert [[Links]] to [Links](Links)
shortlinks: false

# generate a table of contents file with this name, for example:
# toc: Table of Contents
toc: ~

# path to theme folder, for example:
# theme: _theme
theme: ~

# open the server URL in the browser
open: false

# provide user:password for basic authentication, for example:
# auth: admin:s3cr3t
auth: false

# if auth is enabled, specify auth realm name
auth_zone: Restricted Documentation

# show files with these extensions in the navigation and search, for example:
# expose_extensions: pdf,docx,xlsx,txt
expose_extensions: ~

# exclude directories that match these regular expressions
# note that this is an array
exclude: ['^[a-z_\\-0-9]+$']"

"{" + ($contents |
    split row "\n" |
    skip 2 |
    str join $"\n" |
    str replace --multiline --all '# +(.*)\n(#.*\n)*([^:]+): (.+)' '"$3": { "description": "$1", "type": "string", "examples": ["$4"] }' |
    split row "\n" |
    find --invert --regex '^$' |
    str join ",\n") + "}"
```

Output may require some modifications:

```json
{
    "path": {
        "description": "path to the documentation root",
        "type": "string",
        "examples": [
            "."
        ]
    },
    "port": {
        "description": "server port",
        "type": "string",
        "examples": [
            "3000"
        ]
    },
    "bind": {
        "description": "server listen address",
        "type": "string",
        "examples": [
            "0.0.0.0"
        ]
    },
    "base_uri": {
        "description": "set a server root path, for example:",
        "type": "string",
        "examples": [
            "~"
        ]
    },
    "sidebar": {
        "description": "enable sidebar",
        "type": "string",
        "examples": [
            "true"
        ]
    },
    "auto_h1": {
        "description": "add H1 title to files that do not have one",
        "type": "string",
        "examples": [
            "true"
        ]
    },
    "auto_nav": {
        "description": "append navigation to directory READMEs",
        "type": "string",
        "examples": [
            "true"
        ]
    },
    "auto_toc": {
        "description": "replace <!-- TOC --> in any file with its internal table of contents",
        "type": "string",
        "examples": [
            "true"
        ]
    },
    "highlighter": {
        "description": "enable syntax highlighter for code snippets",
        "type": "string",
        "examples": [
            "true"
        ]
    },
    "copy_code": {
        "description": "enable the copy to clipboard icon for code snippets",
        "type": "string",
        "examples": [
            "true"
        ]
    },
    "shortlinks": {
        "description": "convert [[Links]] to [Links](Links)",
        "type": "string",
        "examples": [
            "false"
        ]
    },
    "toc": {
        "description": "generate a table of contents file with this name, for example:",
        "type": "string",
        "examples": [
            "~"
        ]
    },
    "theme": {
        "description": "path to theme folder, for example:",
        "type": "string",
        "examples": [
            "~"
        ]
    },
    "open": {
        "description": "open the server URL in the browser",
        "type": "string",
        "examples": [
            "false"
        ]
    },
    "auth": {
        "description": "provide user:password for basic authentication, for example:",
        "type": "string",
        "examples": [
            "false"
        ]
    },
    "auth_zone": {
        "description": "if auth is enabled, specify auth realm name",
        "type": "string",
        "examples": [
            "Restricted Documentation"
        ]
    },
    "expose_extensions": {
        "description": "show files with these extensions in the navigation and search, for example:",
        "type": "string",
        "examples": [
            "~"
        ]
    },
    "exclude": {
        "description": "exclude directories that match these regular expressions",
        "type": "string",
        "examples": ["['^[a-z_\\-0-9]+$']"
        ]
    }
}
```

DannyBen commented:
Nice. But then the actual schema is more verbose than this initial output.

I wonder if there isn't already a CLI that converts any file (YAML, JSON) with structured comments, to schema. Shouldn't be difficult to create one if there isn't.

Although - I think I mentioned this earlier - I think that if you create the JSON schemas in YAML, and just convert YAML to JSON with jq, the source will be much nicer to look at and work with. JSON is yucky.... :)
